DRAFT - feature-extraction /  feature-matching.mdx (#106)
johko/computer-vision-course: chapters/en/Unit 1 - Fundamentals/feature-extraction/feature-matching.mdx

## Brute-Force Search

Imagine you have a giant box of puzzle pieces, and you're trying to find a specific piece that fits into your puzzle. This is similar to searching for matching features in images. Instead of having any special strategy, you decide to check every piece, one by one until you find the right one. This straightforward method is a brute-force search. The advantage of brute force is its simplicity. You don't need any special tricks – just patience. However, it can be time-consuming, especially if there are a lot of pieces to check. In the context of feature matching, this brute force approach is akin to comparing every pixel in one image to every pixel in another to see if they match. It's exhaustive and it might take a lot of time, especially for large images.

Now that we have an intuitive idea of how brute-force matches are found, let's dive into the algorithms. We are going to use the descriptors that we learned about in the previous chapter to find the matching features in two images.

First install and load libraries

```bash
!pip install opencv-python
```

```python
import cv2
import numpy as np
```

> **Brute Force with SIFT**

Reviewer: Amazing tutorial! I only have one *small* suggestion. In my opinion, using a heading (e.g. Heading 3 in this case) instead of bold for the algorithms would make it much easier for navigation.

Reviewer: After checking the history, I found that you were using heading 3 for the steps and @merveenoyan suggested using bold instead. For this to not feel redundant😅, let me be a little more clear. I'm just suggesting using the heading for the algorithms and not the steps. Again, this is just a suggestion, your writing is truly wonderful!

**We will now load our images to have their features matched.**

```python
img1 = cv.imread('image1.png',cv.IMREAD_GRAYSCALE)
img2 = cv.imread('image2.png',cv.IMREAD_GRAYSCALE)
```

> **Initiatize SIFT detector**

Reviewer 2: ```suggestion
**Initialize SIFT detector**
```

```python
sift = cv.SIFT_create()
```

**Find the keypoints and descriptors with SIFT**

```python
kp1, des1 = sift.detectAndCompute(img1,None)
kp2, des2 = sift.detectAndCompute(img2,None)
```

**Find matches using k nearest neighbors**

```python
bf = cv.BFMatcher()
matches = bf.knnMatch(des1,des2,k=2)
```

**Apply ratio test to threshold the best matches**

```python
good = []
for m,n in matches:
    if m.distance < 0.75*n.distance:
        good.append([m])
```

**Draw the matches**

```python
img3 = cv.drawMatchesKnn(img1,kp1,img2,kp2,good,None,flags=cv.DrawMatchesFlags_NOT_DRAW_SINGLE_POINTS)
```

![SIFT](URL)

**Brute Force with ORB (binary) descriptors**

**Load images**

```python
    img1 = cv.imread('image1.png',cv.IMREAD_GRAYSCALE)
    img2 = cv.imread('image2.png',cv.IMREAD_GRAYSCALE)
```

**Initialize the ORB descriptor**

```python
orb = cv.ORB_create()
```

**Find keypoints and descriptors**

```python
kp1, des1 = orb.detectAndCompute(img1,None)
kp2, des2 = orb.detectAndCompute(img2,None)
```

Because ORB is a binary descriptor, we find matches using [Hamming Distance](https://www.geeksforgeeks.org/hamming-distance-two-strings/),

> which is a measure of the difference between two strings of equal length

Reviewer 2: ```suggestion
which is a measure of the difference between two strings of equal length.
```

# Feature Matching

> How can we match detected features from image-to-image? Feature matching involves comparing key features or characteristics in different

Reviewer 3: ```suggestion
How can we match detected features from one image to another? Feature matching involves comparing key features or characteristics in different
```

> images to find similarities. For example, you might want to match features between two images to identify objects or recognize patterns.

Reviewer 3: this section continues in new line

Author: I have rewritten this and hopefully fixed the line-break issue.

## Brute-Force Search

Imagine you have a giant box of puzzle pieces, and you're trying to find a specific piece that fits
into your puzzle. This is similar to searching for matching features in images. Instead of having any special strategy,

> you decide to check every single piece, one by one, until you find the right one. This straightforward method is like a brute force search.

Reviewer 3: ```suggestion
you decide to check every piece, one by one until you find the right one. This straightforward method is a brute-force search.
```

The advantage of brute force is its simplicity. You don't need any special tricks – just patience. However, it can be time-consuming,
especially if there are a lot of pieces to check. In the context of feature matching, this brute force approach is akin to comparing every pixel in one image to every pixel
in another to see if they match. It's exhaustive and it might take a lot of time, especially for large images.

> OK, now that we have an intuitive idea of how brute force matches are found, let's dive into the algorithms.

Reviewer 3: ```suggestion
Now that we have an intuitive idea of how brute-force matches are found, let's dive into the algorithms.
```

> _First install and load libraries_

Reviewer 3: ```suggestion
First install and import the libraries.
```

```bash
!pip install opencv-python
```

```python
import cv2
import numpy as np
```

> _Brute Force with SIFT_

Reviewer 3: ```suggestion
**Brute Force with SIFT**
```
maybe you could define what SIFT does here

Author: I didn't provide a discussion  b/c it is covered in feature-description.md.  Do you think I should add something here as well?

> 1.  Load Images

Reviewer 3: ```suggestion
We will now load our images to have their features matched.
```

```python
img1 = cv.imread('image1.png',cv.IMREAD_GRAYSCALE)
img2 = cv.imread('image2.png',cv.IMREAD_GRAYSCALE)
```

> 2. Initiate SIFT detector

Reviewer 3: ```suggestion
Initialize SIFT detector.
```

```python
sift = cv.SIFT_create()
```

> 3. Find the keypoints and descriptors with SIFT

Reviewer 3: ```suggestion
Find the keypoints and descriptors with SIFT.
```

```python
kp1, des1 = sift.detectAndCompute(img1,None)
kp2, des2 = sift.detectAndCompute(img2,None)
```

> 4. Find matches using k nearest neighbors

Reviewer 3: ```suggestion
Find matches using k-nearest neighbors.
```

```python
bf = cv.BFMatcher()
matches = bf.knnMatch(des1,des2,k=2)
```

> 5. Apply ratio test to threshold the best matches

Reviewer 3: ```suggestion
Apply ratio test to threshold the best matches.
```

```python
good = []
for m,n in matches:
    if m.distance < 0.75*n.distance:
        good.append([m])
```

> 5. Draw the matches

Reviewer 3: ```suggestion
Draw the matches.
```

```python
img3 = cv.drawMatchesKnn(img1,kp1,img2,kp2,good,None,flags=cv.DrawMatchesFlags_NOT_DRAW_SINGLE_POINTS)
```

![SIFT](URL)

> _Brute Force with ORB (binary) descriptors_

Reviewer 3: ```suggestion
**Brute Force with ORB (binary) descriptors**
```
You can also have them level three headers (###)

> 1.  Load images

Reviewer 3: ```suggestion
Load images.
```

```python
    img1 = cv.imread('image1.png',cv.IMREAD_GRAYSCALE)
    img2 = cv.imread('image2.png',cv.IMREAD_GRAYSCALE)
```

> 2.  Intialize the ORB descriptor:

Reviewer 3: ```suggestion
Initialize the ORB descriptor.
```

```python
orb = cv.ORB_create()
```

> 4. Find key points and descriptors:

Reviewer 3: ```suggestion
Find key points and descriptors.
```

```python
kp1, des1 = orb.detectAndCompute(img1,None)
kp2, des2 = orb.detectAndCompute(img2,None)
```

> 3.  Because ORB is a binary descriptor, we ind matches using Hamming Distance, which is a measure of the difference between

Reviewer 3: ```suggestion
Because ORB is a binary descriptor, we find matches using Hamming Distance, which is a measure of the difference between
```
Maybe you could give a link to hamming distance

    two strings of equal length

```python
bf = cv.BFMatcher(cv.NORM_HAMMING, crossCheck=True)
```

> 4.  Find the matches

Reviewer 3: ```suggestion
Find the matches.
```

```python
matches = bf.match(des1,des2)
```

> 5. Sort them in the order of their distance.

Reviewer 3: ```suggestion
Sort them in the order of their distance.
```

```python
matches = sorted(matches, key = lambda x:x.distance)
```

> 6. Draw first n matches.

Reviewer 3: ```suggestion
Draw the first n matches.
```

This means it can make educated guesses about which pieces might fit well together, even if they're not an exact match. Under the hood, FLANN is uses something called k-D trees.
Think of it as organizing the puzzle pieces in a special way. Instead of checking every piece against every other piece, FLANN arranges them in a tree-like structure that makes finding matches faster.
In each node of the k-D tree, FLANN puts pieces with similar features together. It's like sorting puzzle pieces with similar shapes or colors into piles.
This way, when you're looking for a match, you can quickly check the pile that's most likely to have similar pieces. Let's say you're looking for a "sky" piece.
Instead of searching through all the pieces, FLANN guides you to the right spot in the k-D tree where the sky-colored pieces are sotred. FLANN also adjusts its strategy based on the features of the puzzle pieces.
If you have a puzzle with lots of colors, it will focus on color features. Alternately, if it's a puzzle with intricate shapes, it pays attention to those shapes.
By balancing speed and accuracy when finding matching features, FLANN substantially improves query time.

> _First we create a dictionary to specify the algorithm we will use:_

Reviewer 3: ```suggestion
First we create a dictionary to specify the algorithm we will use.
```

- For SIFT or SURF

```python
FLANN_INDEX_KDTREE = 1
index_params = dict(algorithm = FLANN_INDEX_KDTREE, trees = 5)
```

- For ORB

```python
FLANN_INDEX_LSH = 6
index_params= dict(algorithm = FLANN_INDEX_LSH,
                   table_number = 12,
                   key_size = 20,
                   multi_probe_level = 2)
```

_We also create a dictionary to specify the maximum leafs to visit:_

```python
search_params = dict(checks=50)
```

> 1.  Load images

Reviewer 3: ```suggestion
Load images.
```

````python
img1 = cv.imread('image1.png',cv.IMREAD_GRAYSCALE)
img2 = cv.imread('image2.png',cv.IMREAD_GRAYSCALE)

> 2. Initiate SIFT detector

Reviewer 3: ```suggestion
Initialize SIFT detector.
```

```python
sift = cv.SIFT_create()

> ````

Reviewer 3: ```suggestion
```
```

> 3. Find the keypoints and descriptors with SIFT

Reviewer 3: ```suggestion
Find the keypoints and descriptors with SIFT.
```

```python
kp1, des1 = sift.detectAndCompute(img1,None)
kp2, des2 = sift.detectAndCompute(img2,None)
```

> 3. Define the FLANN parameters. Trees is the numbers of bins you want.

Reviewer 3: ```suggestion
Define the FLANN parameters. Trees is the numbers of bins you want.
```

```python
FLANN_INDEX_KDTREE = 1
index_params = dict(algorithm = FLANN_INDEX_KDTREE, trees = 5)
search_params = dict(checks=50)
flann = cv.FlannBasedMatcher(index_params,search_params)

matches = flann.knnMatch(des1,des2,k=2)
```

> 4. We will only draw only good matches, so create a mask

Reviewer 3: ```suggestion
We will only draw only good matches, so create a mask.
```

```python
matchesMask = [[0,0] for i in range(len(matches))]
```

> 5. We perform a ratio test to determine good matches

Reviewer 3: ```suggestion
We perform a ratio test to determine good matches.
```

```python
for i,(m,n) in enumerate(matches):
    if m.distance < 0.7*n.distance:
        matchesMask[i]=[1,0]
```

> 6. Now we visualize the matches

Reviewer 3: ```suggestion
Now we visualize the matches.
```

As we have already learned, it is important that the matching algorithm handles changes in rotation or scale. If a feature is turned or resized, LoFTR would still recognize it. It's like solving puzzles where pieces may be flipped or adjusted.
As LoFTR matches features, it assigns a similarity score to indicate how well the features align. Higher scores mean better matches. It's like giving a grade to how well one puzzle piece fits with another.
LoFTR is also invariant to certain transformations, meaning it can handle variations in lighting, angle, or perspective. This is crucial when dealing with images that might be photographed under different conditions.
LoFTR's ability to robustly match features makes it valuable for tasks like image stitching, where you combine multiple images seamlessly by identifying and connecting common features.

We can use [Kornia](https://github.com/kornia/kornia) to find matching features in two images using LoFTR.

```bash
!pip install kornia  kornia-rs  kornia_moons opencv-python --upgrade
```

> 1.  Import the libraries

Reviewer 3: ```suggestion
Import the libraries.
```

```python
import cv2
import kornia as K
import kornia.feature as KF
import matplotlib.pyplot as plt
import numpy as np
import torch
from kornia_moons.viz import draw_LAF_matches
```

> 2. Load and resize the images

Reviewer 3: ```suggestion
Load and resize the images.
```

```python
from kornia.feature import LoFTR

img1 = K.io.load_image(image1.jpg, K.io.ImageLoadType.RGB32)[None, ...]
img2 = K.io.load_image(image2.jpg, K.io.ImageLoadType.RGB32)[None, ...]

img1 = K.geometry.resize(img1, (512, 512), antialias=True)
img2 = K.geometry.resize(img2, (512, 512), antialias=True)
```

> 3.  Indicate whether the image is an "indoor" or "outdoor" image

Reviewer 3: ```suggestion
Indicate whether the image is an "indoor" or "outdoor" image.
```

```python
matcher = LoFTR(pretrained="indoor")
```

> 4.  LoFTR only works on grayscale images, so convert to grayscale

Reviewer 3: ```suggestion
LoFTR only works on grayscale images, so convert to grayscale.
```

```python
input_dict = {
    "image0": K.color.rgb_to_grayscale(img1),
    "image1": K.color.rgb_to_grayscale(img2),
}
```

> 5. Perform Inference

Reviewer 3: ```suggestion
Perform inference.
```

```python
with torch.inference_mode():
    correspondences = matcher(input_dict)
```

> 6. We can clean up the correspondences using Random Sample Consensus (RANSAC), which help

Reviewer 3: ```suggestion
We can clean up the correspondences using Random Sample Consensus (RANSAC), which help
```

   deal with noise or outliers in the data.

```python
mkpts0 = correspondences["keypoints0"].cpu().numpy()
mkpts1 = correspondences["keypoints1"].cpu().numpy()
Fm, inliers = cv2.findFundamentalMat(mkpts0, mkpts1, cv2.USAC_MAGSAC, 0.5, 0.999, 100000)
inliers = inliers > 0
```

> 7. Finally, we visualize the matches

Reviewer 3: ```suggestion
Finally, we visualize the matches
```

```python
draw_LAF_matches(
    KF.laf_from_center_scale_ori(
        torch.from_numpy(mkpts0).view(1, -1, 2),
        torch.ones(mkpts0.shape[0]).view(1, -1, 1, 1),
        torch.ones(mkpts0.shape[0]).view(1, -1, 1),
    ),
    KF.laf_from_center_scale_ori(
        torch.from_numpy(mkpts1).view(1, -1, 2),
        torch.ones(mkpts1.shape[0]).view(1, -1, 1, 1),
        torch.ones(mkpts1.shape[0]).view(1, -1, 1),
    ),
    torch.arange(mkpts0.shape[0]).view(-1, 1).repeat(1, 2),
    K.tensor_to_image(img1),
    K.tensor_to_image(img2),
    inliers,
    draw_dict={"inlier_color": (0.2, 1, 0.2), "tentative_color": None, "feature_color": (0.2, 0.5, 1), "vertical": False},
)
```

The best matches are visualized in green, while less certain matches are in blue.

![LoFTR](URL)

## Resources and Further Reading

> [FLANN Github](https://github.com/flann-lib/flann)

Reviewer 3: Maybe have them as a list with - for readability?

## Brute-Force Search

Imagine you have a giant box of puzzle pieces, and you're trying to find a specific piece that fits
into your puzzle. This is similar to searching for matching features in images. Instead of having any special strategy,
you decide to check every piece, one by one until you find the right one. This straightforward method is a brute-force search.
The advantage of brute force is its simplicity. You don't need any special tricks – just patience. However, it can be time-consuming,
especially if there are a lot of pieces to check. In the context of feature matching, this brute force approach is akin to comparing every pixel in one image to every pixel
in another to see if they match. It's exhaustive and it might take a lot of time, especially for large images.

Now that we have an intuitive idea of how brute-force matches are found, let's dive into the algorithms. We are going to use the
descriptors that we learned about in the previous chapter to find the matching features in two images.

First install and load libraries

```bash
!pip install opencv-python
```

```python
import cv2
import numpy as np
```

**Brute Force with SIFT**

> ### We will now load our images to have their features matched.

Reviewer 3: This will appear smaller than **Brute Force with SIFT** which is just a bold text, maybe for each step above code snippets you can turn into bold instead of making each a header

```python
img1 = cv.imread('image1.png',cv.IMREAD_GRAYSCALE)
img2 = cv.imread('image2.png',cv.IMREAD_GRAYSCALE)
```

> ### Initiatize SIFT detector

Reviewer 4: Do the learners know at this point what a SIFT detector does/is?

Author: This should be covered in previous chapter (feature description)

```python
sift = cv.SIFT_create()
```

> ### Find the keypoints and descriptors with SIFT

Reviewer 4: Again, are the learners familiar with what the keypoints and descriptors are?

Author: feature description.mdx discusses the SIFT algorithm and keypoints/descriptors -- let me know if I need to add more detail here

```python
kp1, des1 = sift.detectAndCompute(img1,None)
kp2, des2 = sift.detectAndCompute(img2,None)
```

### Find matches using k nearest neighbors

```python
bf = cv.BFMatcher()
matches = bf.knnMatch(des1,des2,k=2)
```

### Apply ratio test to threshold the best matches

```python
good = []
for m,n in matches:
    if m.distance < 0.75*n.distance:
        good.append([m])
```

### Draw the matches

```python
img3 = cv.drawMatchesKnn(img1,kp1,img2,kp2,good,None,flags=cv.DrawMatchesFlags_NOT_DRAW_SINGLE_POINTS)
```

![SIFT](URL)

**Brute Force with ORB (binary) descriptors**

### Load images

```python
    img1 = cv.imread('image1.png',cv.IMREAD_GRAYSCALE)
    img2 = cv.imread('image2.png',cv.IMREAD_GRAYSCALE)
```

> ### Intialize the ORB descriptor:

Reviewer 3: ```suggestion
### Initialize the ORB descriptor:
```

```python
orb = cv.ORB_create()
```

> ### Find key points and descriptors:

Reviewer 4: Earlier, it was spelled "keypoints", here it is "key points". Which one is correct? Also, a colon at the end of a header is unnecessary.

```python
kp1, des1 = orb.detectAndCompute(img1,None)
kp2, des2 = orb.detectAndCompute(img2,None)
```

> ### Because ORB is a binary descriptor, we find matches using [Hamming Distance](https://www.geeksforgeeks.org/hamming-distance-two-strings/), which is a measure of the difference between

Reviewer 4: This probably shouldn't be a header, but a regular paragraph.

> two strings of equal length

Reviewer 3: Because of newline this is not formatted

Author: I think  I have removed the newlines

```python
bf = cv.BFMatcher(cv.NORM_HAMMING, crossCheck=True)
```

### Find the matches

```python
matches = bf.match(des1,des2)
```

> ### Sort them in the order of their distance.

Reviewer 4: ```suggestion
### Sort them in the order of their distance
```

```python
matches = sorted(matches, key = lambda x:x.distance)
```

> ### Draw first n matches.

Reviewer 4: ```suggestion
### Draw first n matches
```

This means it can make educated guesses about which pieces might fit well together, even if they're not an exact match. Under the hood, FLANN is uses something called k-D trees.
Think of it as organizing the puzzle pieces in a special way. Instead of checking every piece against every other piece, FLANN arranges them in a tree-like structure that makes finding matches faster.
In each node of the k-D tree, FLANN puts pieces with similar features together. It's like sorting puzzle pieces with similar shapes or colors into piles.
This way, when you're looking for a match, you can quickly check the pile that's most likely to have similar pieces. Let's say you're looking for a "sky" piece.
Instead of searching through all the pieces, FLANN guides you to the right spot in the k-D tree where the sky-colored pieces are sotred. FLANN also adjusts its strategy based on the features of the puzzle pieces.
If you have a puzzle with lots of colors, it will focus on color features. Alternately, if it's a puzzle with intricate shapes, it pays attention to those shapes.
By balancing speed and accuracy when finding matching features, FLANN substantially improves query time.

> ### First we create a dictionary to specify the algorithm we will use.

Reviewer 4: ```suggestion
### First we create a dictionary to specify the algorithm we will use
```

**For SIFT or SURF**

```python
FLANN_INDEX_KDTREE = 1
index_params = dict(algorithm = FLANN_INDEX_KDTREE, trees = 5)
```

**For ORB**

We will use the parameters from the paper.

```python
FLANN_INDEX_LSH = 6
index_params= dict(algorithm = FLANN_INDEX_LSH,
                   table_number = 12,
                   key_size = 20,
                   multi_probe_level = 2)
```

> _We also create a dictionary to specify the maximum leafs to visit:_

Reviewer 4: ```suggestion
We also create a dictionary to specify the maximum leafs to visit:
```

```python
search_params = dict(checks=50)
```

Load images

```python
img1 = cv.imread('image1.png',cv.IMREAD_GRAYSCALE)
img2 = cv.imread('image2.png',cv.IMREAD_GRAYSCALE)
```

Initiate SIFT detector

```python
sift = cv.SIFT_create()
```

> ### Find the keypoints and descriptors with SIFT.

Reviewer 4: ```suggestion
### Find the keypoints and descriptors with SIFT
```

```python
kp1, des1 = sift.detectAndCompute(img1,None)
kp2, des2 = sift.detectAndCompute(img2,None)
```

### Define the FLANN parameters. Trees is the numbers of bins you want.

```python
FLANN_INDEX_KDTREE = 1
index_params = dict(algorithm = FLANN_INDEX_KDTREE, trees = 5)
search_params = dict(checks=50)
flann = cv.FlannBasedMatcher(index_params,search_params)

matches = flann.knnMatch(des1,des2,k=2)
```

> ### We will only draw only good matches, so create a mask

Reviewer 4: ```suggestion
### We will only draw good matches, so create a mask
```
